### PR TITLE
Add authorization middleware based on roles

### PR DESCRIPTION
**Pull request by Cartmanishere**
On backend, some of the routes are only allowed to be called for certain sets of users.
For e.g
- Super user (a.k.a @ankushdharkar )
- App owners (as defined by the list of mockdata in the repo)

Instead of writing different middleware for authorization of each particular type of user, 
this PR aims to provide a unified way of setting authorization for any route.

This is based on the user object having `roles` information.

A user object will contain the roles information as follows:
```
{
  roles: {
    super_user: true/false,
            OR
    app_owner: true/false
  }
}
```

There is hierarchy of required roles. For e.g `superUser` has higher authority than `appOwner`. 
So any route which requires `appOwner` permission, can also be done by a `superUser`.

But a user with `app_owner` role cannot do a request which requires authorization of `superUser`.

### CODE REVIEW

File: middlewares/authorization.js [hunk 1]
```diff
@@ -1,13 +1,56 @@
...
+ * for that route. Currently following roles are supported:
+ * - `authorizeUser('superUser')`
+ * - `authorizeUser('appOwner')`
+ * Note: This must be added on routes after the `authenticate` middleware.
+ * @param {String} requiredRole - The least role authority required for a route.
+ * @returns {Function} - A middleware function that authorizes given role.
+ */
+const authorizeUser = (requiredRole) => {
+  return (req, res, next) => {
+    let { roles } = req.userData
```
Comment (ankurnarkhede) [inline middlewares/authorization.js:40]: Add default as `{}` if roles is `undefined`.

**Review by ankurnarkhede — CHANGES_REQUESTED** [review]

File: middlewares/authorization.js [hunk 2]
```diff
@@ -1,13 +1,56 @@
...
+const REQUIRED_ROLES_PRIORITY = {
+  superUser: ['super_user'],
+  appOwner: ['app_owner', 'super_user'],
+  default: ['default', 'super_user', 'app_owner']
+}
+
+/**
+ * Check if the user has enough authorization based on their role.
+ * User would have following roles schema -
+ * userRoles = {member: true, app_owner: true, super_user: true}
```
Comment (Cartmanishere) [inline middlewares/authorization.js:17]: `member` is not used. update comment.

File: firebase.json [hunk 3]
```diff
@@ -1,7 +1,7 @@
 {
   "emulators": {
     "firestore": {
-      "port": 8080
+      "port": 8081
```
Comment (ankurnarkhede) [inline firebase.json:4]: This change can be reverted, @Cartmanishere!

File: routes/stocks.js [hunk 4]
```diff
@@ -62,6 +62,6 @@ router.get('/', fetchStocks)
  *           schema:
  *             $ref: '#/components/schemas/errors/badImplementation'
  */
-router.post('/', authenticate, authorization, createStock, addNewStock)
+router.post('/', authenticate, authorizeUser('superUser'), createStock, addNewStock)
```
Comment (ankurnarkhede) [inline routes/stocks.js:65]: Can we have the strings `'superUser', 'appOwner', etc` defined as a constant and then use it here. We might mess up with the strings sometimes leading to an issue!

**Review by ankurnarkhede — CHANGES_REQUESTED** [review]
@Cartmanishere , added few comments.
Let me know if you can complete the roles strings to be defined as a constant by 1st May. Else, we can merge this PR and track it for later in an another issue.

**Review by ankurnarkhede — APPROVED** [review]
LGTM ✅ 
To be merged after constants for roles are added or the corresponding issue is created.

**Comment by Cartmanishere** [comment]
@ankurnarkhede Created an issue for the changes you have requested.

I will pick this up in sometime or if someone wants, they can work on it too.

https://github.com/Real-Dev-Squad/website-backend/issues/353

**Pull request merged**